packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,196 @@
+    )
+
+    const disposeOnPatches = onPatches(subtreeRoot, patches => {
+      if (this.withDraftPatchRecorder) {
+        throw failure("original subtree must not change while 'withDraft' executes")
+      }
+      this.allowWrite(() => {
+        applyPatches(this.subtreeRootDraft, patches)
+      })
+    })
+
+    const { allowWrite, dispose: disposeReadonlyMW } = readonlyMiddleware(this.subtreeRootDraft)
+    this.allowWrite = allowWrite
+
+    this.disposer = () => {
+      disposeReactionRS()
+      disposeOnPatches()
+      disposeReadonlyMW()
+      if (isRootStore(this.subtreeRootDraft)) {
+        unregisterRootStore(this.subtreeRootDraft)
+      }
+      this.disposer = () => {}
+    }
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @typeparam R Return type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is called with a draft copy of `node`. Any changes made to the draft
+   * are applied to the original subtree when `fn` returns `true` or `{ commit: true, ... }`. When
+   * `fn` returns `false` or `{ commit: false, ... }` the changes made to the draft are rejected.
+   * @returns Value of type `R` when `fn` returns an object of type `{ commit: boolean; return: R }`
+   * or `void` when `fn` returns a boolean.
+   */
+  withDraft<T extends object, R = void>(node: T, fn: WithDraftCallback<T, R>): R {
+    assertIsFunction(fn, "fn")
+
+    const { draftNode, applyRecorderChanges } = this.prepareDraftChanges(node)
+
+    let commit = false
+    try {
+      const returnValue = this.allowWrite(() => fn(draftNode))
+      if (typeof returnValue === "boolean") {
+        commit = returnValue
+        return undefined as any
+      } else {
+        commit = returnValue.commit
+        return returnValue.return
+      }
+    } finally {
+      applyRecorderChanges(commit)
+    }
+  }
+
+  /**
+   * Disposes the draft middleware.
+   */
+  dispose(): void {
+    this.disposer()
+  }
+
+  private prepareDraftChanges<T extends object>(
+    node: T
+  ): { draftNode: T; applyRecorderChanges: (commit: boolean) => void } {
+    assertTweakedObject(node, "node")
+
+    const isDraft = !!this.withDraftPatchRecorder
+
+    const path = getParentToChildPath(isDraft ? this.subtreeRootDraft : this.subtreeRoot, node)
+    if (!path) {
+      throw failure(`node is not a child of subtreeRoot${isDraft ? "Draft" : ""}`)
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      throw failure("path could not be resolved - draft may be out of sync with original tree")
+    }
+
+    if (!this.withDraftPatchRecorder) {
+      this.withDraftPatchRecorder = patchRecorder(this.subtreeRootDraft)
+    }
+    const recorder = this.withDraftPatchRecorder
+    const numRecorderEvents = recorder.events.length
```

Comment:
ah, so you only rollback the amount of events that the inner call generated, smart!

Reply:
Yes, thanks 🙂